guides/recipes/recursive-jobs.md

```diff
@@ -0,0 +1,106 @@
+# Recursive Jobs
+
+Recursive jobs, like recursive functions, call themselves after they have have
+executed. Except unlike recursive functions, where recursion happens in a tight
+loop, a recursive job enqueues a new version of itself and may add a slight
+delay to alleviate pressure on the queue.
+
+Recursive jobs are a great way to backfill large amounts of data where a
+database migration or a mix task may not be suitable. Here are a few reasons
+that a recursive job may be better suited for backfilling data:
+
+* Data can't be backfilled with a database migration, it may require talking to
+  an external service
+* A task may fail partway through execution; resuming the task would mean
+  starting over again, or tracking progress manually to resume where the failure
+  occurred
+* A task may be computationally intensive or put heavy pressure on the database
+* A task may run for too long and would be interrupted by code releases or other
+  node restarts
+* A task may interface with an external service and require some rate limiting
+* A job can be used directly for new records _and_ to backfill existing records
+
+Let's explore recursive jobs with a use case that builds on several of those
+reasons.
+
+## Use Case: Backfilling Timezone Data
+
+Consider a worker that queries an external service to determine what timezone a
+user resides in. The external service has a rate limit and the response time is
+unpredictable. We have a lot of users in our database missing timezone
+information, and we need to backfill.
+
+Our application has an existing `TimezoneWorker` that accepts a user's `id`,
+makes an external request and then updates the user's timezone. We can modify
+the worker to handle backfilling by adding a new clause to `perform/1`. The new
+clause explicitly checks for a `backfill` argument and will enqueue the next job
+after it executes:
+
+```elixir
+defmodule MyApp.Workers.TimezoneWorker do
+  use Oban.Worker
+
+  import Ecto.Query
+
+  alias MyApp.{Repo, User}
+
+  @backfill_delay 1
+
+  @impl true
+  def perform(%{args: %{"id" => id, "backfill" => true}}) do
+    with :ok <- perform(%{args: %{"id" => id}}),
+         next_id when is_integer(next_id) <- fetch_next(id) do
```

Comment:
Here we could get into an issue because `fetch_next/1` can return `nil` (let's say we just processed the last user with `timezone = nil`), then this function would return `nil`, and that would break the contract.

Reply:
The contract for `perform` was much looser back then. I think a `case` or an `if` would be better to explicitly control the flow.

Reply:
Yes, I added a `case` after my self-review :)